### PR TITLE
Nouns Explorer

### PR DESCRIPTION
**Pull request by ripe0x**
## TLDR
a new Explore page that shows all nouns in a single view, and also allows visitors to see the details of each noun’s traits.

### key functionality:

- grid of all existing Nouns
- descending & ascending date sorting
- Noun detail view with a list of each trait used
- galaxy view showing a zoomed-out view of the grid
- desktop
    - sticky sidebar on scroll with selected Noun’s details
    - hover to quickly navigate Nouns
    - ← → ↑ ↓ keyboard navigation
- mobile
    - modal view for Noun details
    - swipe left/right on touch screens to quickly navigate Noun details without returning to grid view

### Context
Discourse thread: https://discourse.nouns.wtf/t/small-grants-add-nouns-explorer-to-nouns-wtf/2164

image
image
image

### nouns-assets

- adds `getPartData` to `packages/nouns-assets/src/utils.ts` to get encoded part data for a single trait

### nouns-sdk

- adds `buildPartSVG` to `packages/nouns-sdk/src/image/svg-builder.ts` to get create an svg image from the part data

### nouns-webapp

- adds packages: framer motion, react-swipeable
- modifies global styles in `App.module.css` to only apply overflow-x on small screens, which resolves a conflict with `position: fixed` and the sticky sidebar
- adds hook `useKeyPress` to enable keyboard navigation on Explore page
- NavBar
    - adds new component: `NavDropdown` enabling desktop nav items to expand and reveal child nav links
    - modifies `NavBarButton` to allow for a nav button to allow for icon + arrow in button
- Nouns + Traits
    - adds new function `StandaloneNounImage` to `StandaloneNoun` to fetch a noun image without a link wrapper.
    - adds new component `StandalonePart` to fetch an individual part image
- Explore page
    - adds new page and route `Explore`
    - adds new component `ExploreGrid` to fetch, sort, and navigate Noun data
    - adds new component `ExploreGridItem` to display images and handle errors with fallbacks
    - adds new component `ExploreNav` to handle grid reordering
    - adds new component `ExploreNounDetail` to display Noun details, re-order and format trait data from default, and navigate to other nouns

## Testing

- confirmed key functionality works as expected on desktop and mobile
- tested in chome, firefox, safari, brave, and ios and android devices

### CODE REVIEW

File: packages/nouns-assets/src/utils.ts [hunk 1]
```diff
@@ -82,3 +82,15 @@ export const getNounSeedFromBlockHash = (nounId: BigNumberish, blockHash: string
...
+ * Get encoded part information for one trait
+ * @param partType The label of the part type to use
+ * @param partIndex The index within the image data array of the part to get
+ */
+ export const getPartData = (partType:string, partIndex:number) => {
+  type ObjectKey = keyof typeof images;
+  const part = partType as ObjectKey;
+  return images[part][partIndex].data;
+};
+exports.getPartData = getPartData;
```
Comment (solimander) [inline packages/nouns-assets/src/utils.ts:96]: `exports.getPartData = getPartData;` is unnecessary

File: packages/nouns-sdk/src/image/svg-builder.ts [hunk 2]
```diff
@@ -84,3 +85,48 @@ export const buildSVG = (
 
   return `${svgWithoutEndTag}</svg>`;
 };
+
+/**
+ * Given the RLE data of an individual part, palette colors, and a background color, build an SVG image.
+ * @param partData The RLE part datas
+ * @param paletteColors The hex palette colors
+ */
+export const buildPartSVG = (
```
Comment (solimander) [inline packages/nouns-sdk/src/image/svg-builder.ts:94]: Nit: should be able to accomplish the same thing with `buildSVG`, right? Maybe just make `bgColor` optional and skip the background fill if it's not passed.

File: packages/nouns-sdk/src/image/svg-builder.ts [hunk 3]
```diff
@@ -54,6 +54,7 @@ export const buildSVG = (
     let currentY = bounds.top;
 
     rects.forEach(draw => {
+      // eslint-disable-next-line prefer-const
```
Comment (solimander) [inline packages/nouns-sdk/src/image/svg-builder.ts:57]: Nit: Fix rather than silencing error. e.g.

```typescript
let drawLength = draw[0];
const colorIndex = draw[1];
```

File: packages/nouns-webapp/src/components/ExploreGrid/ExploreNounDetail/index.tsx [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+import { AnimatePresence, motion } from 'framer-motion/dist/framer-motion';
+import { XIcon } from '@heroicons/react/solid';
+import NounInfoRowBirthday from '../../NounInfoRowBirthday';
+import loadingNoun from '../../../assets/loading-skull-noun.gif';
+import Placeholder from 'react-bootstrap/Placeholder';
+import Image from 'react-bootstrap/Image';
+import cx from 'classnames';
+import { useSwipeable } from 'react-swipeable';
+import dotenv from 'dotenv';
+dotenv.config();
```
Comment (solimander) [inline packages/nouns-webapp/src/components/ExploreGrid/ExploreNounDetail/index.tsx:17]: Already run in `index.ts`

File: packages/nouns-webapp/src/components/ExploreGrid/index.tsx [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        const start = i - rangeChunkSize < 0 ? 0 : i - rangeChunkSize;
+        const end = i - 1;
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars
+        const nounsRange = await fetchNouns(start, end);
+      }
+    }
+  };
+
+  const fetchNouns = async (start: number, end: number) => {
+    const url = `https://noun.pics/range?start=${start}&end=${end}`;
```
Comment (solimander) [inline packages/nouns-webapp/src/components/ExploreGrid/index.tsx:71]: These external calls could be avoided as we already have enough local information to construct existing Nouns.

That is, the cached Nouns seeds, which are indexed by Noun ID, and the RLE `ImageData`.

File: packages/nouns-webapp/src/pages/Explore/index.tsx [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+import cx from 'classnames';
+import ExploreNounDetail from '../../components/ExploreGrid/ExploreNounDetail';
+import { motion, AnimatePresence } from 'framer-motion/dist/framer-motion';
+import { Auction as IAuction } from '../../wrappers/nounsAuction';
+import { useAppSelector } from '../../hooks';
+import dotenv from 'dotenv';
+import { useKeyPress } from '../../hooks/useKeyPress';
+import ExploreNav from '../../components/ExploreGrid/ExploreNav';
+import ExploreGrid from '../../components/ExploreGrid';
+dotenv.config();
```
Comment (solimander) [inline packages/nouns-webapp/src/pages/Explore/index.tsx:13]: Already run in `index.ts`

**Comment by lastpunk9999** [comment]
On iOS Safari, when visiting the [deploy preview](https://deploy-preview-595--nouns-testnet.netlify.app/), then clicking `Nouns & Traits`, the Explore page loads displays Nouns starting at Id 498 instead of the current Noun Id 505

IMG_36103804D486-1

File: packages/nouns-sdk/src/image/svg-builder.ts (hunk 2 above)
Comment (ripe0x) [reply packages/nouns-sdk/src/image/svg-builder.ts:94]: `buildSVG` accepts and iterates through the part data of multiple traits. and this new function accepts individual part data. 
they may be able to be combined, but i believe it'll require some additional changes beyond the `bgColor` conditional, and i didn't want to mess with this critical function too much. but i'm happy to combine them into if that makes more sense.

File: packages/nouns-assets/src/utils.ts (hunk 1 above)
Comment (ripe0x) [reply packages/nouns-assets/src/utils.ts]: resolved in https://github.com/nounsDAO/nouns-monorepo/pull/595/commits/89fcb7e23c0444fb453b48388036134070c513a0

File: packages/nouns-sdk/src/image/svg-builder.ts (hunk 3 above)
Comment (ripe0x) [reply packages/nouns-sdk/src/image/svg-builder.ts]: resolved in https://github.com/nounsDAO/nouns-monorepo/pull/595/commits/dc4f60b4c037d1eb25377ce8e4986324ee243062

File: packages/nouns-webapp/src/components/ExploreGrid/ExploreNounDetail/index.tsx (hunk 4 above)
Comment (ripe0x) [reply packages/nouns-webapp/src/components/ExploreGrid/ExploreNounDetail/index.tsx]: resolved in https://github.com/nounsDAO/nouns-monorepo/pull/595/commits/26745469ba53750e9b5cdf25f0d94368ea6f02ff

File: packages/nouns-webapp/src/pages/Explore/index.tsx (hunk 6 above)
Comment (ripe0x) [reply packages/nouns-webapp/src/pages/Explore/index.tsx]: resolved in https://github.com/nounsDAO/nouns-monorepo/pull/595/commits/ac419510caf31635bd6269e46d86cdd58e0a7397

File: packages/nouns-webapp/src/components/ExploreGrid/index.tsx (hunk 5 above)
Comment (ripe0x) [reply packages/nouns-webapp/src/components/ExploreGrid/index.tsx:71]: iirc, that was actually the original approach here (using `StandaloneNounImage`) but with so many images on one page it felt slow to interact with, so we opted for static images via noun.pics.

File: packages/nouns-sdk/src/image/svg-builder.ts (hunk 2 above)
Comment (solimander) [reply packages/nouns-sdk/src/image/svg-builder.ts:94]: Yeah, doesn't quite seem necessary to have an entirely different function when you could achieve the same by doing:

```typescript
buildSVG([{ data }], palette);
// rather than
buildPartSVG(data, palette)
```

File: packages/nouns-webapp/src/components/ExploreGrid/index.tsx (hunk 5 above)
Comment (solimander) [reply packages/nouns-webapp/src/components/ExploreGrid/index.tsx:71]: Interesting, wouldn't expect the image generation to add much overhead. Parsing the base64 image data is probably more CPU-intensive, but both solutions are bound by that. Maybe due to `StandaloneNounImage`, which should be trimmed down if using here.

Regardless, can push it off and possibly update later.

**Comment by ripe0x** [comment]
> On iOS Safari, when visiting the [deploy preview](https://deploy-preview-595--nouns-testnet.netlify.app/), then clicking `Nouns & Traits`, the Explore page loads displays Nouns starting at Id 498 instead of the current Noun Id 505. A refresh is required to get the display the current Noun(s).

good catch @lastpunk9999! that's been resolved in https://github.com/nounsDAO/nouns-monorepo/pull/595/commits/b961254bd09d009e1c8ccdb3f4c2dd02fd344ff8. seen in [this preview link](https://frabjous-kitsune-25fe40.netlify.app/explore).

File: packages/nouns-webapp/src/components/ExploreGrid/index.tsx (hunk 5 above)
Comment (ripe0x) [reply packages/nouns-webapp/src/components/ExploreGrid/index.tsx:72]: i'll see if i can dig up a deploy preview link for us to compare against. and in the meantime i'll mark this this issue resolved for now.

File: packages/nouns-sdk/src/image/svg-builder.ts (hunk 2 above)
Comment (ripe0x) [reply packages/nouns-sdk/src/image/svg-builder.ts]: yes! that's a much simpler solution. thanks for the guidance here @solimander.

this has been resolved in https://github.com/nounsDAO/nouns-monorepo/pull/595/commits/61b814dd7e5570898dd9eb1b8f9214586758637d

**Comment by ripe0x** [comment]
thanks for the review @lastpunk9999 & @solimander. ⌐◨-◨

the issues noted have been marked as resolved and the latest commits [can be previewed here](https://636e4d65ffd8a2000939ffd2--frabjous-kitsune-25fe40.netlify.app/explore).

**Review by solimander — APPROVED** [review]
⌐◨-◨
